Analyses/src/CountVDHits_module.cc

```diff
@@ -0,0 +1,86 @@
+// Pawel Plesniak, 2024
+
+#include <string>
+#include <iostream>
+#include <bits/stdc++.h>
+#include <vector>
+#include <algorithm>
+
+#include "art/Framework/Core/EDAnalyzer.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "canvas/Utilities/InputTag.h"
+#include "Offline/MCDataProducts/inc/StepPointMC.hh"
+
+#include "CLHEP/Vector/ThreeVector.h"
+#include "CLHEP/Units/PhysicalConstants.h"
+namespace mu2e {
+
+  //================================================================
+  class CountVDHits : public art::EDAnalyzer {
+  public:
+    explicit CountVDHits(fhicl::ParameterSet const& pset);
+    void analyze(const art::Event& evt) override;
+    void endJob() override;
+  private:
+    art::InputTag _stepPointMCsTag;
+    std::vector<int> _enabledVDs;
+    std::vector<int> _enabledVDStepPointMCsCount;
+  };
+
+  //================================================================
+  CountVDHits::CountVDHits(const fhicl::ParameterSet& pset)
+    : art::EDAnalyzer(pset),
+      _stepPointMCsTag(pset.get<std::string>("StepPointMCsTag")),
+      _enabledVDs(pset.get<std::vector<int> >("enableVDs", std::vector<int>()))
+  {
+    // Sort all elements to be in ascending order
+    std::sort(_enabledVDs.begin(), _enabledVDs.end());
+    // Remove all duplicate elements
+    std::vector<int>::iterator _uniqueIndex = std::unique(_enabledVDs.begin(), _enabledVDs.end());
+    // Removed undefined memory
+    _enabledVDs.resize(std::distance(_enabledVDs.begin(), _uniqueIndex));
+    // Insert _enabledVDs.size() zeros into the counter vector
+    _enabledVDStepPointMCsCount.insert(_enabledVDStepPointMCsCount.end(), _enabledVDs.size(), 0);
+  }
+
+  //================================================================
+  void CountVDHits::analyze(const art::Event& event) {
+    // Get the hits corresponding to the StepPointMCCollection of interest
+    art::Handle<StepPointMCCollection> StepPointMCs;
+    event.getByLabel(_stepPointMCsTag, StepPointMCs);
+    if (!(StepPointMCs.isValid())){return;}
+
+    // Define the index here so there's less overhead with memory allocation
+    int index;
+
+    // Loop over all the StepPointMCs to find which ones cross the VDs of interest. If a StepPointMC does not correspond to a volume ID, it is ignored.
+    for (auto &StepPointMC : *StepPointMCs)
+    {
+      //Find the position off the volume ID
+      auto it = find(_enabledVDs.begin(), _enabledVDs.end(), StepPointMC.volumeId());
+      // If the volumeID is not in the list, skip to the next event
+      if (it == _enabledVDs.end()){continue;}
+      // Increment the counter for the corresponding entry
+      index = it - _enabledVDs.begin();
+      _enabledVDStepPointMCsCount[index]++;
+    }
+    return;
+  }
+
+  //================================================================
+  void CountVDHits::endJob()
+  {
+    std::cout << "Summary " << std::endl;
```

Comment:
I am unsure of how to decorate appropriately or whether there is a defined standard, if you could review the change I have made for this I will amend as necessary

Reply:
There are no real standards. I suggest you start it with \n so that it's separated from what's before it.  My tastes in decorations are different than your's but they are just tastes."

Something to be aware of, there is  a message logger facility that will add decorations like 

```
%MSG-i Category:  module_type:module_label@EndJob 04-Jun-2024 16:27:29 CDT ModuleEndJob
  Your message content goes here
  It may be multiple lines
%MSG
```

If you are interested I can point you to examples and docs.

For this module, a verbose true/false is OK.  In the general case an int verbosity is more useful - higher numbers mean more printout.

Reply:
I would be interested, if you send me some examples and doc I can see what I can implement, get them integrated in ASAP

Reply:
On any mu2e machine, look under:

/exp/mu2e/app/users/kutschke/Development/ForPawel

There is a subdirectory with an example module and fcl.  The key points are:

1. `#include "messagefacility/MessageLogger/MessageLogger.h"`
2. Check that   'MF_MessageLogger', is in the link lists in SConscript
3. You can use LogInfo two ways.  In the analyze member function I don't bother making a named object contruct and write to it in one line.
4. In the EndJob member function I make a named object so I can re-use it.
5. LogInfo ignores std::endl.  It obeys "\n"
6. You don't need the final "\n" because that is added for you.
7. Where I have "RK" you can use "STM".  That's called a category.  Categories are described in:

See https://mu2ewiki.fnal.gov/wiki/MessageLogger

Reply:
Thank you for this Rob, that helped a lot :)